e2e-tests/desktop/tests/targets.spec.js

```diff
@@ -4,19 +4,228 @@
+      .getByLabel('Toggle secret visibility')
+      .click();
+    await expect(
+      authedPage
+        .getByRole('listitem')
+        .filter({ hasText: 'private_key' })
+        .locator('pre'),
+    ).toContainText('BEGIN OPENSSH PRIVATE KEY');
 
+    await authedPage.getByRole('button', { name: 'End Session' }).click();
+    await expect(authedPage.getByText('Canceled successfully.')).toBeVisible();
+    await expect(authedPage).toHaveURL(
+      /.*\/scopes\/global\/projects\/targets$/,
+    );
+  });
+
+  [{ host: 'Quick Connect' }, { host: hostName }].forEach(({ host }) => {
+    test(`Connects to a tcp target with two hosts using ${host}`, async ({
+      authedPage,
+    }) => {
+      await authedPage
+        .getByRole('link', { name: targetWithTwoHosts.name })
+        .click();
+      await authedPage.getByRole('button', { name: 'Connect' }).click();
+
+      await expect(authedPage).toHaveURL(
+        /.*\/scopes\/global\/projects\/targets/,
+      );
+      await authedPage.getByRole('button', { name: host }).click();
+
+      await expect(authedPage).toHaveURL(
+        /.*\/scopes\/global\/projects\/sessions/,
+      );
+
+      await expect(authedPage.getByText(credential.name)).toBeVisible();
+      await expect(authedPage.getByText('username')).toBeVisible();
+      await expect(authedPage.getByText('private_key')).toBeVisible();
+
+      await authedPage
+        .getByRole('listitem')
+        .filter({ hasText: 'username' })
+        .getByLabel('Toggle secret visibility')
+        .click();
+      await expect(
+        authedPage
+          .getByRole('listitem')
+          .filter({ hasText: 'username' })
+          .locator('pre'),
+      ).toHaveText(credential.attributes.username);
+
+      await authedPage
+        .getByRole('listitem')
+        .filter({ hasText: 'private_key' })
+        .getByLabel('Toggle secret visibility')
+        .click();
+      await expect(
+        authedPage
+          .getByRole('listitem')
+          .filter({ hasText: 'private_key' })
+          .locator('pre'),
+      ).toContainText('BEGIN OPENSSH PRIVATE KEY');
+
+      await authedPage.getByRole('button', { name: 'End Session' }).click();
+      await expect(
+        authedPage.getByText('Canceled successfully.'),
+      ).toBeVisible();
+      await expect(authedPage).toHaveURL(
+        /.*\/scopes\/global\/projects\/targets$/,
+      );
+    });
+  });
+
+  test('Connects to an SSH target', async ({ authedPage }) => {
+    await authedPage.getByRole('link', { name: sshTarget.name }).click();
     await authedPage.getByRole('button', { name: 'Connect' }).click();
 
     await expect(authedPage).toHaveURL(
       /.*\/scopes\/global\/projects\/sessions/,
     );
 
+    await authedPage.getByRole('tab', { name: 'Shell' }).click();
+    // TODO: Research a better way to test canvas elements for the shell,
+    //  would it be too brittle to assert a snapshot of an expected image?
```

Comment:
snapshots feel potentially brittle, especially with that session timer and the ephemeral port on that page? 
![Screenshot 2024-11-26 at 11 26 27 AM](https://github.com/user-attachments/assets/8e4f0917-51c5-4f75-8349-09a6317aba9a)

Are we able to get the text contents of the shell? Could we we wait and ensure that the shell gets a shell prompt?

Reply:
Unfortunately you can't just pull the text contents out otherwise I would totally do that. Canvas elements are basically images so there's not much to do with it unless you mock it out

Reply:
🤔  do we have access to these canvas methods? I think mocking these methods (if we have access) is better than visually testing the shell

Reply:
We do not, but either way that would defeat the purpose of E2E tests